### DISCUSSION TITLE
Are b32 addresses public

### DISCUSSION THREAD

**Discussion by arthurmelton** [PurpleI2P/i2pd#2049] [Q&A]
I am wondering if b32 addresses (created specifically by SAM) are announced in a way that an attacker could find newly created b32 addresses. I don't see how it would be possible without publicly telling everyone that a new service is running, but I would rather know for sure.

**Comment by anikey-from-i2p** [comment]
As far as I know (and this is general for all i2p addresses, not just SAM), if you create an address and leave it at all default settings, its existance (or specifically its leaseset) can be visible to floodfill routers. This is because the leaseset is not encrypted (aka not "blinded").

This issue can be solved by using blinded leasesets. From reading [some](https://geti2p.net/en/docs/api/samv3) [documentation](https://geti2p.net/spec/common-structures#key-certificates) (see also [this](https://i2pd.readthedocs.io/en/latest/user-guide/tunnels/)), this can be achieved by setting `SIGNATURE_TYPE` to `11` (which is `RedDSA_SHA512_Ed25519`). I think you also need to set the option `i2cp.leaseSetType` to `5` (which means encrypted; again see the i2pd.readthedocs website i linked).

I've not tested this. Be sure to test it. Java I2P's `NetDB Lookup` is useful in showing data about leasesets (as an aside to developers: why not implement that kinda thing in i2pd webconsole? should i open issue?).

Another option (can be used together with the above) could be to implement an "accesslist" (like i2pd does for server tunnels), where you can put trusted addresses and, on connection, check if it is one of those addresses. But i'm sure that's something you'll have to add in the application side.

**Comment by Vort** [comment]
> Java I2P's NetDB Lookup is useful in showing data about leasesets (as an aside to developers: why not implement that kinda thing in i2pd webconsole? should i open issue?).

Can you show screenshot?
Isn't this is what you are talking about?

![image](https://github.com/PurpleI2P/i2pd/assets/1242858/5575b59c-36b1-4451-ab05-c2affbd6b1f6)

**Reply by anikey-from-i2p** [reply]
It does indeed look something like that. The problem with it is that it appears to be only available to floodfills.

With Java I2P, you can contact any i2p destination through that router, then open router console, NetDB Lookup, and in the `Full destination, name, Base32, or hash:` field enter the destination you visited.

Let's say you want to see leaseset for i2p-projekt.i2p. You open it in your browser (now the router knows the leaseset for it). Then you do NetDB Lookup. And in response you get something like this:
```
LeaseSet: XXXX.... Destination: i2p-projekt.i2p
Published XX sec ago
Expires in X min
RAP? false  RAR? true  Distance: 123.XX  Type: 3  Unpublished? false
Signature type: DSA_SHA1
Encryption Key: ECIES_X25519 XXXX....
Encryption Key: ELGAMAL_2048 XXXX....
Routing Key: XXXX....

    Lease 1: ?? XXXX Tunnel 1234567 Expires in X min
    Lease 2: ?? XXXX Tunnel 1234567 Expires in X min
```

Can i do this in i2pd? Or is that panel you showed, only shows the leasesets stored in the floodfill?

**Reply by Vort** [reply]
> Or is that panel you showed, only shows the leasesets stored in the floodfill?

Yes, **LeaseSets** section appear only in floodfill mode.

> And in response you get something like this:

Ok, I understood what you mean.

It would be nice to have leaseset address clickable here in a similar way it is clickable in **LeaseSets** section:
![image](https://github.com/PurpleI2P/i2pd/assets/1242858/963ea89e-fbe2-4120-8aa6-998312a4762f)

However I doubt developers will make such changes anytime soon - i2pd have lots of more important unsolved problems.
